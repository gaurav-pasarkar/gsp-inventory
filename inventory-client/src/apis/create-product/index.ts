import {Product} from "../../components/page/buy-product";

const createProduct = async (product: Product) => {

  return await new Promise(resolve => {
    // @ts-ignore
    google.script.run.withSuccessHandler(resolve).submitProduct({
      product_name: product.productName,
      cost_price: product.costPrice,
      selling_price: product.sellingPrice,
      quantity: product.quantity
    })
  })
}

export default createProduct