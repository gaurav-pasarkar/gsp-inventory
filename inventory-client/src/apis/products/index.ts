export interface Product {
  productName: string,
  costPrice: number,
  sellingPrice: number,
  quantity: number
}

interface ApiProduct {
  product_name: string,
  cost_price: number,
  selling_price: number,
  quantity: number
}

export const createProduct = async (product: Product) => {

  const apiProduct = {
    product_name: product.productName,
    cost_price: product.costPrice,
    selling_price: product.sellingPrice,
    quantity: product.quantity
  }

  return await new Promise(resolve => {
    // @ts-ignore
    google.script.run.withSuccessHandler(resolve).submitProduct(apiProduct)
  })
}

export const getProducts = async (): Promise<Product[]> => {

  const transformProducts = (products: ApiProduct[]) => products.map(p => ({
    productName: p.product_name,
    costPrice: p.cost_price,
    sellingPrice: p.selling_price,
    quantity: p.quantity
  }))

  return await new Promise(resolve => {
    // @ts-ignore
    google.script.run
      .withSuccessHandler((products: ApiProduct[]) => resolve(transformProducts(products)))
      .getAvailableProducts()
  })
}