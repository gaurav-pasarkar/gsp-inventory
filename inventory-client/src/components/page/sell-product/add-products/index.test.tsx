import matchSnapshot from "../../../../test-utils/match-snapshot";
import AddProducts from "./index";
import {getProducts} from "../../../../apis/products";
import {act} from "react-test-renderer";
import {fireEvent, render, waitFor, within, screen, wait} from "@testing-library/react";

describe('Add product to cart', () => {

  it('should render', async () => {
    const availableProducts = await getProducts();
    matchSnapshot(<AddProducts availableProducts={availableProducts}/>)
   });

   it('should render products', async () => {
     const availableProducts = await getProducts();
     await act(async () => {
       const {getByTestId} = render(<AddProducts availableProducts={availableProducts}/>);

       const autocomplete = getByTestId('productName');
       const input = within(autocomplete).getByRole('textbox')
       autocomplete.focus()
       fireEvent.change(input, { target: { value: 'P' } })
       fireEvent.keyDown(autocomplete, { key: 'ArrowDown' })
       fireEvent.keyDown(autocomplete, { key: 'Enter' })

       await waitFor(() => {
         expect(within(autocomplete).getByRole('textbox')).toHaveValue('Product 1')
       })
     });
   })
})