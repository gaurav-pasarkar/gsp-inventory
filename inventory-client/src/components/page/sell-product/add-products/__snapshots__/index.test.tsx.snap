// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add product to cart should render 1`] = `
Array [
  <div
    aria-controls="panel1d-content"
    aria-disabled={false}
    className="MuiButtonBase-root MuiAccordionSummary-root WithStyles(ForwardRef(AccordionSummary))-root-3"
    id="panel1d-header"
    onBlur={[Function]}
    onClick={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    role="button"
    tabIndex={0}
  >
    <div
      className="MuiAccordionSummary-content WithStyles(ForwardRef(AccordionSummary))-content-4"
    >
      <p
        className="MuiTypography-root MuiTypography-body1"
      >
        Add Products
      </p>
    </div>
  </div>,
  <div
    className="MuiAccordionDetails-root WithStyles(ForwardRef(AccordionDetails))-root-6"
  >
    <form
      className="makeStyles-root-1"
      onSubmit={[Function]}
    >
      <div
        aria-expanded={false}
        aria-owns={null}
        className="MuiAutocomplete-root MuiAutocomplete-fullWidth MuiAutocomplete-hasClearIcon MuiAutocomplete-hasPopupIcon"
        data-testid="productName"
        onClick={[Function]}
        onKeyDown={[Function]}
        onMouseDown={[Function]}
        role="combobox"
      >
        <div
          className="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
        >
          <label
            className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
            data-shrink={false}
            id="undefined-label"
          >
            Product Name
          </label>
          <div
            className="MuiInputBase-root MuiOutlinedInput-root MuiAutocomplete-inputRoot MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiOutlinedInput-adornedEnd"
            onClick={[Function]}
          >
            <input
              aria-activedescendant={null}
              aria-autocomplete="list"
              aria-controls={null}
              aria-invalid={false}
              autoCapitalize="none"
              autoComplete="off"
              autoFocus={false}
              className="MuiInputBase-input MuiOutlinedInput-input MuiAutocomplete-input MuiAutocomplete-inputFocused MuiInputBase-inputAdornedEnd MuiOutlinedInput-inputAdornedEnd"
              disabled={false}
              onAnimationStart={[Function]}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onMouseDown={[Function]}
              required={false}
              spellCheck="false"
              type="text"
              value=""
            />
            <div
              className="MuiAutocomplete-endAdornment"
            >
              <button
                aria-label="Clear"
                className="MuiButtonBase-root MuiIconButton-root MuiAutocomplete-clearIndicator"
                disabled={false}
                onBlur={[Function]}
                onClick={[Function]}
                onDragLeave={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseLeave={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
                tabIndex={-1}
                title="Clear"
                type="button"
              >
                <span
                  className="MuiIconButton-label"
                >
                  <svg
                    aria-hidden={true}
                    className="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                    />
                  </svg>
                </span>
              </button>
              <button
                aria-label="Open"
                className="MuiButtonBase-root MuiIconButton-root MuiAutocomplete-popupIndicator"
                disabled={false}
                onBlur={[Function]}
                onClick={[Function]}
                onDragLeave={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseLeave={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
                tabIndex={-1}
                title="Open"
                type="button"
              >
                <span
                  className="MuiIconButton-label"
                >
                  <svg
                    aria-hidden={true}
                    className="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M7 10l5 5 5-5z"
                    />
                  </svg>
                </span>
              </button>
            </div>
            <fieldset
              aria-hidden={true}
              className="PrivateNotchedOutline-root-7 MuiOutlinedInput-notchedOutline"
            >
              <legend
                className="PrivateNotchedOutline-legendLabelled-9"
              >
                <span>
                  Product Name
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
      <div>
        <button
          className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-containedSizeLarge MuiButton-sizeLarge"
          disabled={false}
          onBlur={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="submit"
        >
          <span
            className="MuiButton-label"
          >
            Add product
          </span>
        </button>
      </div>
    </form>
  </div>,
]
`;
